rfcs/0155-nixos-migrations.md

```diff
@@ -0,0 +1,118 @@
+# Motivation for Migrations 
+
+While the current NixOS Configuration system works incredibly well for immutable services, not all services are immutable, and we currently lack facilities to handling imperative steps needed to upgrade or downgrade complex stateful services. Nix can even be an obstacle to doing so, as often these steps are dissonant with the methodologies required for declarative and immutable structures.
+
+Let's take the example of GitLab. GitLab is notoriously hard to upgrade, and the current NixOS module system produces major obstacles to upgrades of this nature, as GitLab expects the user to run many imperative steps to modify stateful parts of the application such as the database and configuration files. The need to migrate stateful portions of an application to new versions is nothing new; database migrations are standard practice and can provide structure and inform the concerns of module migrations.
+
+With the extension of the NixOS Configuration system with migration infrastructure, we can close this conceptual gap and fully realize the value of atomic immutable deployments, as we gain the ability to atomically migrate parts of stateful services along with the packages and definitions.
+
+# Proposed Migration Structure
+
+The following extension to the module system and standard is proposed. Each migration is a Set defined in Nix with the following fields:
+
+- `version`
+- `up`
+- `down`
+
+`up` and `down` are both Nix Sets with the following fields:
+
+- `script`
+- `warn`
+- `backup`
+- `restore`
+
+## Version Information
+
+For migrations to work we must know what version the stateful parts of this service are currently using; this will be stored in the metadata of the systemd service in a standard location to be determined. This is needed for the same reason as in the database migration context: to determine which migrations are needed to achieve the new state.
+
+```Nix
+{
+  config.migrations = [ { version = "2023-04-07"; ... } ];
+}
+```
+
+The version information will be stored as a formatted date string (such that lexigraphical and chronological sort are the same), representing the date at which the migration was merged into Nixpkgs.
+
+## Up Script
+
+A place in the Nix Module system for describing the imperative steps required to migrate from the previous version to the new version. This will be run as a systemd "oneshot" service, to take advantage of the standard architecture.
```

Comment:
I would prefer if migrations were not fully automatically triggered, for example for nixos-unstable users it could be important to make a backup and supervise the process.